src/util.ts

```diff
@@ -53,3 +60,62 @@ export function when<T, R>(condition: any, fn: (x: T) => R) {
     return condition ? fn(x) : x
   }
 }
+
+/**
+ * Simple finite state machine with explicit allowed state transitions,
+ * initial state, and options around how long to wait for state transitions
+ * when using waitForChange as well as explicitly naming the machine for
+ * better debugging when using multiple machines simultaneously.
+ */
+export function fsm<T extends string>(
+  stateTransitions: StateTransitions<T>,
+  initState: T,
+  options: Options & FsmOptions<T> = {}
+) {
+  let state = initState
+  const name = options.name || 'fsm'
+  const onStateChange = options.onStateChange
+
+  const is = (...states: T[]) => states.includes(state)
+
+  const change = (newState: T) => {
+    debug('%s changing state from %s to %s', name, state, newState)
+    if (!stateTransitions[state]?.includes(newState)) {
+      throw new StateError(
+        `${name} invalid state transition - ${state} to ${newState}`
+      )
+    }
+    const oldState = state
+    state = newState
+    debug('%s changed state from %s to %s', name, oldState, newState)
+    if (onStateChange) {
+      onStateChange({ name, from: oldState, to: newState })
+    }
+  }
+
+  const waitForChange = (...newStates: T[]) => {
+    debug(
+      '%s waiting for state change from %s to %s',
+      name,
+      state,
+      newStates.join(' or ')
+    )
+    return waitUntil(() => newStates.includes(state), options)
+  }
+
+  return {
+    /**
+     * Transition state to `newState`. Throws if `newState` is not a valid
+     * transitional state for the current state.
+     */
+    change,
+    /**
+     * Wait for state to change to one of `newStates`. Times out after 5 seconds.
```

Comment:
Where is the 5 second timeout coming from here? I am not sure I see anything passing in a parameter for that

Reply:
That's the default behavior for the `async-wait-until` package. It can be overriden. BTW, I pulled this code out into it's own package - https://github.com/smartprocure/simple-machines